src/Shared/UrlDecoder/UrlDecoder.cs

```diff
@@ -345,5 +345,291 @@ private static bool SkipUnescape(int value1, int value2, bool isFormEncoding)
+                {
+                    return false;
+                }
+
+                var nextSourceIndex = sourceIndex;
+                var nextChar = UnescapePercentEncoding(ref nextSourceIndex, buffer);
+                if (nextChar == -1)
+                {
+                    return false;
+                }
+
+                // When UnescapePercentEncoding returns -1 we shall return false.
+                // For performance reasons, there is no separate if statement for the above check
+                // as the condition below also returns -1 for that case.
+                if ((nextChar & 0xC0) != 0x80)
+                {
+                    // the follow up char is not in form of 10xx xxxx
+                    return false;
+                }
+
+                currentDecodeBits = (currentDecodeBits << 6) | (nextChar & 0x3F);
+                remainingChars--;
+
+                sourceIndex = nextSourceIndex;
+            }
+
+            if (currentDecodeBits < expectValueMin)
+            {
+                // overlong encoding
+                return false;
+            }
+
+            if (!System.Text.Rune.TryCreate(currentDecodeBits, out var rune) || !rune.TryEncodeToUtf16(buffer.Slice(destinationIndex), out var charsWritten))
+            {
+                // Reasons for this failure could be:
+                // Value is in the range of 0xD800-0xDFFF UTF-16 surrogates that are not allowed in UTF-8
+                // Value is above the upper Unicode bound of 0x10FFFF
+                return false;
+            }
+
+            destinationIndex += charsWritten;
+            return true;
+        }
+
+        /// <summary>
+        /// Read the percent-encoding and try unescape it.
+        ///
+        /// The operation first peek at the character the <paramref name="scan"/>
+        /// iterator points at. If it is % the <paramref name="scan"/> is then
+        /// moved on to scan the following to characters. If the two following
+        /// characters are hexadecimal literals they will be unescaped and the
+        /// value will be returned.
+        ///
+        /// If the first character is not % the <paramref name="scan"/> iterator
+        /// will be removed beyond the location of % and -1 will be returned.
+        ///
+        /// If the following two characters can't be successfully unescaped the
+        /// <paramref name="scan"/> iterator will be move behind the % and -1
+        /// will be returned.
+        /// </summary>
+        /// <param name="scan">The value to read</param>
+        /// <param name="buffer">The char array</param>
+        /// <returns>The unescaped char if success. Otherwise return -1.</returns>
+        private static int UnescapePercentEncoding(ref int scan, ReadOnlySpan<char> buffer)
+        {
+            int tempIdx = scan;
+            if (buffer[tempIdx] != '%')
+            {
+                return -1;
+            }
+            scan = tempIdx + 1;
```

Comment:
I think my transformation is incorrect here.

Reply:
I think you intended something like:

```cs
int probe = scan++;
if (buffer[probe] != '%') { return -1; }
int firstNibble = /* remainder of method */;
```